### PR TITLE
JUnit v4 -> v5 Migration

### PR DESCRIPTION
**Pull request by ham1**
- `Ignored` -> `Disabled`.
- `@Test(expected ...)` -> `Assertions.assertThrows(...)`
- A few `Assert` -> `Assertions`
  - These will be completed in another PR
- Removed error collector.
- Some whitespace.

### CODE REVIEW

**Comment by vlsi** [comment]
What is your suggestion to invite new contributions to use JUnit 5 rather than 4?

File: src/core/src/test/java/org/apache/jmeter/threads/TestUnmodifiableJMeterVariables.java [hunk 1]
```diff
@@ -55,34 +56,52 @@ public void testGetIteration() {
...
     }
 
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void testIncIteration() {
-        unmodifiables.incIteration();
+        Assertions.assertThrows(
+                UnsupportedOperationException.class,
+                unmodifiables::incIteration
+        );
```
Comment (vlsi) [inline src/core/src/test/java/org/apache/jmeter/threads/TestUnmodifiableJMeterVariables.java]: It might be that `assertThrowsUnsupported(lambda)` could be helpful here so we don't repeat the same `Assertions.assertThrows` every time, but we use just `assertThrowsUnsupported(unmodifiables::incIteration)`, `assertThrowsUnsupported(() -> unmodifiables.remove("some.key"))` and so on

Comment (ham1) [reply src/core/src/test/java/org/apache/jmeter/threads/TestUnmodifiableJMeterVariables.java]: Good idea

**Comment by ham1** [comment]
I prefer Spock, because I find it often more expressive and convenient (e.g. accessing private methods).
But if you had to use JUnit then it's good for the lambda support and generally more sensible: https://developer.ibm.com/dwblog/2017/top-five-reasons-to-use-junit-5-java/

**Comment by vlsi** [comment]
> accessing private methods

which is a code smell anyway, isn't it?

**Comment by vlsi** [comment]
> But if you had to use JUnit then it's good for the lambda support and generally more sensible:

It looks like my intention was not well understood.

The current build script allows writing new tests with JUnit4 style.
I think it would be great if it either completely denied JUnit4 (with clear message like "plz use JUnit5") or suggested the refactoring (e.g. a set of spotless rules over test sources that convert junit annotations and so on).

**Comment by ham1** [comment]
> > accessing private methods
> 
> which is a code smell anyway, isn't it?

Correct; most of the code here hasn't been TDD'd and it feels too much effort for such large restructuring of code so the smells go away?
I find it useful for enabling testing for some of the code we already have.
I think it's better than adding `@deprecated` to a method just for testing.

**Comment by ham1** [comment]
> > But if you had to use JUnit then it's good for the lambda support and generally more sensible:
> 
> It looks like my intention was not well understood.
> 
> The current build script allows writing new tests with JUnit4 style.
> I think it would be great if it either completely denied JUnit4 (with clear message like "plz use JUnit5") or suggested the refactoring (e.g. a set of spotless rules over test sources that convert junit annotations and so on).

How about a check somewhere, checkstyle?, for `import org.unit.Test/Before/After/Assert/Assume/Rule`

**Comment by vlsi** [comment]
>I find it useful for enabling testing for some of the code we already have.

Tests that peek onto the implementation details might become hard to maintain.

>I think it's better than adding @deprecated to a method just for testing.

I'm about to add https://github.com/apiguardian-team/apiguardian so we could do things like `@API(status = INTERNAL, since = "5.2.0")`

**Comment by vlsi** [comment]
> How about a check somewhere, checkstyle?,

checkstyle is bad because its messages are not actionable.

I suggest two options:
a) Spotless. It is nice for cases when it can auto-correct (e.g. suggest a proper test package, and it might be even enough).
b) https://github.com/policeman-tools/forbidden-apis

**Comment by ham1** [comment]
> > How about a check somewhere, checkstyle?,
> 
> checkstyle is bad because its messages are not actionable.

True, however based on my experience so far, you can only (easily) auto-correct Test/Before/After etc. Assertions and Rules etc. require more complexity. So I'm not sure if it's doing only some in Spotless or just banning an import RegEx with a message in CheckStyle or indeed Spotless?

> 
> I suggest two options:
> a) Spotless. It is nice for cases when it can auto-correct (e.g. suggest a proper test package, and it might be even enough).
> b) https://github.com/policeman-tools/forbidden-apis

I'm not sure I like the idea of yet more libraries and code, I think I prefer adding the simple cases to Spotless so it can correct the majority but also some documentation somewhere obvious...

**Comment by vlsi** [comment]
>  you can only (easily) auto-correct Test/Before/After etc.

This PR contains such cases, when simple package update was enough.

> also some documentation somewhere obvious

I don't suggest to implement a tool that automatically converts any JUnit4 code to JUnit5.
The idea is the tool should automatically convert known items, and it should try to print meaningful messages where it can't. Of course, certain cases will be out of scope, but I think it would help people to get familiar with the style.

**Comment by vlsi** [comment]
It looks like this PR reduces the code coverage. @ham1 , have you analyzed why is that?
For instance, it says `src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/sampler/ResourcesDownloader.java` is no longer covered.

**Comment by ham1** [comment]
> It looks like this PR reduces the code coverage. @ham1 , have you analyzed why is that?
> For instance, it says `src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/sampler/ResourcesDownloader.java` is no longer covered.

I had a quick look but I don't understand because there isn't and hasn't been a test for it before.

**Comment by pmouawad** [comment]
Wasn't it covered through integration tests ?

**Comment by vlsi** [comment]
>Wasn't it covered through integration tests ?

@pmouawad , I've added `throw exception`, and it shows that ResourceDownloader was covered in org.apache.jmeter.protocol.http.sampler.ParallelResourcesAndIpSource:
https://travis-ci.org/vlsi/jmeter/jobs/593595044#L1346-L1373

@ham1 , could you try adding the same `if (true) { throw ...}` to your branch so we see how it goes?

**Comment by ham1** [comment]
Ah, now I know why I disabled the test in `ParallelResourcesAndIpSource` because `org.apache.jmeter.protocol.http.sampler.ParallelResourcesAndIpSource > test(String, WireMockServer, JMeterVariables)[1] FAILED` always fails on my machine when trying to connect to wiremock.

Too many threads? Even running `./gradlew :src:protocol:http:test` fails all the time.

```
 org.opentest4j.AssertionFailedError: expected: <url: http://wiremock/index.html
    response: OK
    data.size: 85
    data: <html><body><img src='image1.png'><img src='image2.png'><img src='image3.png'></body>
    - url: http://wiremock/image1.png
      response: OK
      data.size: 8
      data: content1
    - url: http://wiremock/image2.png
      response: OK
      data.size: 8
      data: content2
    - url: http://wiremock/image3.png
      response: OK
      data.size: 8
      data: content3
    - url: http://wiremock/index.html
      response: OK
      data.size: 85
      data: <html><body><img src='image1.png'><img src='image2.png'><img src='image3.png'></body>
    > but was: <url: http://wiremock/index.html
    response: Non HTTP response message: Connect to localhost:33099 [localhost/127.0.0.1] failed: Network is unreachable (connect failed)
    data.size: 13705
    data: org.apache.http.conn.HttpHostConnectException: Connect to localhost:33099 [localhost/127.0.0.1] failed: Network is unreachable (connect failed)
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:156)
        at org.apache.jmeter.protocol.http.sampler.HTTPHC4Impl$JMeterDefaultHttpClientConnectionOperator.connect(HTTPHC4Impl.java:326)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:374)
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:393)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
        at org.apache.jmeter.protocol.http.sampler.HTTPHC4Impl.executeRequest(HTTPHC4Impl.java:850)
        at org.apache.jmeter.protocol.http.sampler.HTTPHC4Impl.sample(HTTPHC4Impl.java:561)
        at org.apache.jmeter.protocol.http.sampler.HTTPSamplerProxy.sample(HTTPSamplerProxy.java:67)
        at org.apache.jmeter.protocol.http.sampler.HTTPSamplerBase.sample(HTTPSamplerBase.java:1282)
        at org.apache.jmeter.protocol.http.sampler.ParallelResourcesAndIpSource.test(ParallelResourcesAndIpSource.java:107)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
```

**Comment by vlsi** [comment]
@ham1 , can you please rebase and verify if `ParallelResourcesAndIpSource` works now?

**Comment by ham1** [comment]
Works for me now 😄

File: src/jorphan/src/test/java/org/apache/jorphan/exec/TestKeyToolUtils.java [hunk 2]
```diff
@@ -41,8 +41,8 @@
     private static String password;
     private static final int validity = 1;
 
-    @BeforeAll
-    public static void setup(@TempDir Path keystoreDir) throws IOException {
+    @BeforeEach
+    public void setup(@TempDir Path keystoreDir) throws IOException {
         keystore = keystoreDir.resolve("dummy-keystore.jks").toFile();
         password = JOrphanUtils.generateRandomAlphanumericPassword(32);
```
Comment (vlsi) [inline src/jorphan/src/test/java/org/apache/jorphan/exec/TestKeyToolUtils.java]: @ham1 , why is this change?
you are initializing static fields in individual tests.

Please, please, refrain from pushing unrelated changes under a single commit.
Here you replace "junit5->juni5" code under "junit4->junit5" name. That is super-confuse those who review the change.


/cc @pmouawad

Comment (ham1) [reply src/jorphan/src/test/java/org/apache/jorphan/exec/TestKeyToolUtils.java]: I don't remember this change, given its single thread and still has static fields, it doesn't look a required change.

Comment (ham1) [reply src/jorphan/src/test/java/org/apache/jorphan/exec/TestKeyToolUtils.java]: What might have happened is some global find/replace broke this but I fixed it incorrectly and didn't notice it.
